Documentation/virt/kvm/kvmi.rst

```diff
@@ -1819,20 +1829,33 @@ sent to the introspection tool. The *CONTINUE* action will set the ``new_val``.
 	};
 
 This event is sent when a hypervisor page fault occurs due to a failed
-permission check in the shadow page tables, the introspection has been
-enabled for this event (see *KVMI_VCPU_CONTROL_EVENTS*) and the event was
-generated for a page in which the introspection tool has shown interest
-(ie. has previously touched it by adjusting the spte permissions).
+permission checks, the introspection has been enabled for this event
+(see *KVMI_VCPU_CONTROL_EVENTS*) and the event was generated for a
+page in which the introspection tool has shown interest (ie. has
+previously touched it by adjusting the spte permissions; see
+*KVMI_VM_SET_PAGE_ACCESS*).
 
-The shadow page tables can be used by the introspection tool to guarantee
+These permissions can be used by the introspection tool to guarantee
 the purpose of code areas inside the guest (code, rodata, stack, heap
 etc.) Each attempt at an operation unfitting for a certain memory
 range (eg. execute code in heap) triggers a page fault and gives the
 introspection tool the chance to audit the code attempting the operation.
 
-``kvmi_event``, guest virtual address (or 0xffffffff/UNMAPPED_GVA),
-guest physical address and the access flags (eg. KVMI_PAGE_ACCESS_R)
-are sent to the introspection tool.
+``kvmi_event`` (with the vCPU state), guest virtual address (or
+-1/UNMAPPED_GVA), guest physical address and the access flags
+(eg. KVMI_PAGE_ACCESS_R) are sent to the introspection tool.
+
+In case of a restricted read access, the guest address is the location
```

Comment:
Maybe being specific about the struct field we are talking about would help ?

the guest address (`kvmi_event_pf.gva`) ...

You could also add info about `regs.rip`:
In this context, the `kvmi_event->event.common.arch.regs.rip` contains the guest vrtual address of the instruction responsible for the memory read / write

What do you think ?

Reply:
Sure.